sys/eepreg/eepreg.c

```diff
@@ -0,0 +1,515 @@
+    }
+}
+
+int eepreg_add(uint32_t *pos, const char *name, uint32_t len)
+{
+    int ret = eepreg_check();
+    if (ret == -ENOENT) {
+        /* reg does not exist, so make a new one */
+        eepreg_reset();
+    }
+    else if (ret < 0) {
+        DEBUG("[eepreg_add] eepreg_check failed\n");
+        return ret;
+    }
+
+    uint32_t reg_end = _get_reg_end();
+
+    uint32_t meta_loc = _get_meta_loc(name);
+
+    if (meta_loc == (uint32_t)UINT_MAX) {
+        /* entry does not exist, so make a new one */
+
+        /* location of the new data */
+        *pos = _get_last_loc(reg_end) - len;
+
+        if (_new_entry(name, len) < 0) {
+            DEBUG("[eepreg_add] not enough space for %s\n", name);
+            return -ENOSPC;
+        }
+
+        return 0;
+    }
+
+    *pos = _get_data_loc(meta_loc, _get_meta_len(meta_loc));
+
+    if (len != _get_data_len(meta_loc, *pos)) {
+        DEBUG("[eepreg_add] %s already exists with different length\n", name);
+        return -EADDRINUSE;
+    }
+
+    return 0;
+}
+
+int eepreg_read(uint32_t *pos, const char *name)
+{
+    int ret = eepreg_check();
+    if (ret < 0) {
+        DEBUG("[eepreg_read] eepreg_check failed\n");
+        return ret;
+    }
+
+    uint32_t meta_loc = _get_meta_loc(name);
+
+    if (meta_loc == (uint32_t)UINT_MAX) {
+        DEBUG("[eepreg_read] no entry for %s\n", name);
+        return -ENOENT;
+    }
+
+    *pos = _get_data_loc(meta_loc, _get_meta_len(meta_loc));
+
+    return 0;
+}
+
+int eepreg_write(uint32_t *pos, const char *name, uint32_t len)
+{
+    uint32_t reg_end = _get_reg_end();
+
+    int ret = eepreg_check();
+    if (ret == -ENOENT) {
+        /* reg does not exist, so make a new one */
+        eepreg_reset();
+    }
+    else if (ret < 0) {
+        DEBUG("[eepreg_write] eepreg_check failed\n");
+        return ret;
+    }
+
+    /* location of the new data */
+    *pos = _get_last_loc(reg_end) - len;
+
+    if (_new_entry(name, len) < 0) {
+        DEBUG("[eepreg_write] not enough space for %s\n", name);
+        return -ENOSPC;
+    }
+
+    return 0;
+}
+
+int eepreg_rm(const char *name)
+{
+    int ret = eepreg_check();
+    if (ret < 0) {
+        DEBUG("[eepreg_rm] eepreg_check failed\n");
+        return ret;
+    }
+
+    uint32_t meta_loc = _get_meta_loc(name);
+
+    if (meta_loc == (uint32_t)UINT_MAX) {
+        DEBUG("[eepreg_rm] no entry for %s\n", name);
+        return -ENOENT;
+    }
+
+    uint32_t reg_end = _get_reg_end();
+    uint32_t last_loc = _get_last_loc(reg_end);
+
+    uint8_t meta_len = _get_meta_len(meta_loc);
+    uint32_t tot_meta_len = reg_end - meta_loc;
+
+    uint32_t data_loc = _get_data_loc(meta_loc, meta_len);
+    uint32_t data_len = _get_data_len(meta_loc, data_loc);
+
+    /* data_loc is above last_loc due to descending order */
+    uint32_t tot_data_len = data_loc - last_loc;
+
+    /* don't allow interrupts when editing registry */
+    unsigned int irq_state = irq_disable();
```

Comment:
I think this could be moved in a more central place: in the EEPROM peripheral driver, `eeprom_write` function. At least for AVR. For STM32s, this is more complex.

Reply:
I have removed the IRQ calls. It was probably not necessary anyway, since you would have to be doing weird things in an interrupt to mess it up.

Reply:
I meant that there should be no problems in eepreg itself with being interrupted, so long as the reads/writes eventually complete. "Weird" things in this case would be manipulating the registry inside of interrupt code.